react/features/lobby/actions.js

```diff
@@ -0,0 +1,176 @@
+// @flow
+
+import React from 'react';
+import { type Dispatch } from 'redux';
+
+import { appNavigate, maybeRedirectToWelcomePage } from '../app';
+import { conferenceLeft, conferenceWillJoin, getCurrentConference } from '../base/conference';
+import { openDialog } from '../base/dialog';
+import { getLocalParticipant } from '../base/participants';
+import { hashCode } from '../base/util';
+import { hideNotification, showNotification, NOTIFICATION_TYPE } from '../notifications';
+
+import { SET_KNOCKING_STATE, SET_LOBBY_MODE_ENABLED } from './actionTypes';
+import { EnableLobbyModeDialog, KnockNotification, LobbyScreen } from './components';
+import logger from './logger';
+
+declare var APP: Object;
+
+/**
+ * Cancels the ongoing knocking and abandones the join flow.
+ *
+ * @returns {Function}
+ */
+export function cancelKnocking() {
+    return (dispatch: Dispatch<any>, getState: Function) => {
+        if (typeof APP !== 'undefined') {
+            // when we are redirecting the library should handle any
+            // unload and clean of the connection.
+            APP.API.notifyReadyToClose();
+            dispatch(maybeRedirectToWelcomePage());
+
+            return;
+        }
+
+        dispatch(conferenceLeft(getCurrentConference(getState)));
+        dispatch(appNavigate(undefined));
+    };
+}
+
+/**
+ * Action to be dispatched when a knocking poarticipant leaves before any response.
+ *
+ * @param {string} id - The ID of the participant.
+ * @returns {hideNotification}
+ */
+export function knockingParticipantLeft(id: string) {
+    return hideNotification(hashCode(id));
+}
+
+/**
+ * Action to set the knocking state of the participant.
+ *
+ * @param {boolean} knocking - The new state.
+ * @returns {{
+ *     state: boolean,
+ *     type: SET_KNOCKING_STATE
+ * }}
+ */
+export function setKnockingState(knocking: boolean) {
+    return {
+        knocking,
+        type: SET_KNOCKING_STATE
+    };
+}
+
+/**
+ * Starts knocking and waiting for approval.
+ *
+ * @param {string} password - The password to bypass knocking, if any.
+ * @returns {Function}
+ */
+export function startKnocking(password?: string) {
+    return (dispatch: Dispatch<any>, getState: Function) => {
+        const { membersOnly } = getState()['features/base/conference'];
+        const localParticipant = getLocalParticipant(getState());
+
+        dispatch(setKnockingState(true));
+        dispatch(conferenceWillJoin(membersOnly));
```

Comment:
won't wee get willJoin twice, when someone lets us in?

Reply:
Is that a problem?

Reply:
I think we won't because at this point the conference will join flow already got interrupted by the lobby screen. But yeah, this is how room lock works too